core/blockchain.go

```diff
@@ -2106,6 +2106,14 @@ func (bc *BlockChain) SetChainHead(head *types.Block) error {
 	}
 	bc.writeHeadBlock(head)
 
+	// Re-execute the reorged chain in case the head state is missing.
+	// It should be done before firing events.
+	if !bc.HasState(head.Root()) {
+		if err := bc.recoverAncestors(head); err != nil {
+			return err
```

Comment:
If we wind up here, we have already done `bc.writeHeadBlock(head)`. Isn't that a problem? And wouldn't it be better if we do this check/recovery a bit earlier in this method?

Reply:
You are right, otherwise we might mess up the RPC, fixed it.